cmd/cloudflared/config/model.go

```diff
@@ -25,11 +27,12 @@ type Tunnel struct {
 
 // DNSResolver represents a client side DNS resolver
 type DNSResolver struct {
-	Enabled    bool     `json:"enabled"`
-	Address    string   `json:"address,omitempty"`
-	Port       uint16   `json:"port,omitempty"`
-	Upstreams  []string `json:"upstreams,omitempty"`
-	Bootstraps []string `json:"bootstraps,omitempty"`
+	Enabled                bool     `json:"enabled"`
+	Address                string   `json:"address,omitempty"`
+	Port                   uint16   `json:"port,omitempty"`
+	Upstreams              []string `json:"upstreams,omitempty"`
+	Bootstraps             []string `json:"bootstraps,omitempty"`
+	MaxUpstreamConnections int      `json:"max_upstream_connections,omitempty"`
```

Comment:
I wonder if this setting belongs here: https://github.com/cloudflare/cloudflared/blob/master/cmd/cloudflared/tunnel/cmd.go#L985 instead.

Reply:
Do we really need this to be configurable? How would the users know how to set this? Can we pick a reasonable default and go with it? Say 100 connections or even 1000?

Reply:
I like your point. We could set that default here: https://github.com/cloudflare/cloudflared/pull/290/commits/c2f51021199b57f7d7f104526208b52763dc6984#diff-6f1cb455303a0da3ac60ed8b8cdcb845a5edb82939f56e2e05939cb625522a76R27 and still keep it override-able if anyone should choose to want a different number (even 0).

Reply:
We should probably never let the user choose 0 as it will crash their machine by eating 100% CPU, so yeah, maybe Igor's right and this doesn't need to be configurable.

Reply:
I think having it configurable makes sense. Say Raspberry Pi may see 100% CPU with 10 concurrent connections. But an enterprise server may easily handle 50 connections.

If there is concern with unlimited upper limit, maybe set the min as 1 and disallow setting it to 0?

As for the default value, it should be something that won't trigger throttling with most DoH providers. Say if the provider has a limit of 5 and the default max is 10, in the worst case, the provide may just reject requests from all 10 connections while cloudflared keeps re-creating 10 connections in the pool.

Reply:
Thanks all for your comments on this. I will review your suggestions this evening and update the PR. I do agree providing a default value would save the user the trial-and-error process involved in tweaking it to their hardware. For me, a value of 5 upstream connections gives me really good results on a RPi 1 Model B, but it may be possible to increase it if necessary on multi-core devices.

Reply:
Default of 5 sounds good, at least we know it works for some users. And let's keep it configurable, it should be able to run well on any hardware.